src/Sylius/Component/Promotion/Generator/CouponGenerator.php

```diff
@@ -27,58 +30,79 @@ class CouponGenerator implements CouponGeneratorInterface
     protected $couponFactory;
 
     /**
-     * @var RepositoryInterface
+     * @var CouponRepositoryInterface
      */
-    protected $repository;
+    protected $couponRepository;
 
     /**
-     * @var EntityManagerInterface
+     * @var ObjectManager
      */
-    protected $manager;
+    protected $objectManager;
+
+    /**
+     * @var GenerationPolicyInterface
+     */
+    protected $generationPolicy;
+
+    /**
+     * @var CouponInterface[]
+     */
+    protected $generatedCoupons = [];
 
     /**
      * @param FactoryInterface $couponFactory
-     * @param RepositoryInterface $repository
-     * @param EntityManagerInterface $manager
+     * @param CouponRepositoryInterface $couponRepository
+     * @param ObjectManager $objectManager
+     * @param GenerationPolicyInterface $generationPolicy
      */
-    public function __construct(FactoryInterface $couponFactory, RepositoryInterface $repository, EntityManagerInterface $manager)
-    {
+    public function __construct(
+        FactoryInterface $couponFactory,
+        CouponRepositoryInterface $couponRepository,
+        ObjectManager $objectManager,
+        GenerationPolicyInterface $generationPolicy
+    ) {
         $this->couponFactory = $couponFactory;
-        $this->repository = $repository;
-        $this->manager = $manager;
+        $this->couponRepository = $couponRepository;
+        $this->objectManager = $objectManager;
+        $this->generationPolicy = $generationPolicy;
     }
 
     /**
      * {@inheritdoc}
      */
-    public function generate(PromotionInterface $promotion, Instruction $instruction)
+    public function generate(PromotionInterface $promotion, InstructionInterface $instruction)
     {
-        $generatedCoupons = [];
+        $this->assertGenerationIsPossible($instruction);
         for ($i = 0, $amount = $instruction->getAmount(); $i < $amount; ++$i) {
+            $code = $this->generateUniqueCode($instruction->getCodeLength());
             $coupon = $this->couponFactory->createNew();
             $coupon->setPromotion($promotion);
-            $coupon->setCode($this->generateUniqueCode());
+            $coupon->setCode($code);
             $coupon->setUsageLimit($instruction->getUsageLimit());
             $coupon->setExpiresAt($instruction->getExpiresAt());
 
-            $generatedCoupons[] = $coupon;
+            $this->generatedCoupons[$code] = true;
 
-            $this->manager->persist($coupon);
+            $this->objectManager->persist($coupon);
         }
 
-        $this->manager->flush();
-
-        return $generatedCoupons;
+        $this->objectManager->flush();
+        $this->generatedCoupons = [];
```

Comment:
BC break here (and it broke our app).

Any reason why you decided to remove the generated coupons?
We explicitly added this with a PR a while ago because it made sense that if you ask to generate something that it would be returned.

What was the reason for it? Is this due to concerns over the quantity of coupons that could be generated? If so I don't think it makes any difference because they're registered with Doctrine now so unless you clear the entity manager you're not saving anything.

If there's no reason for removing I'd like to add it back in so let me know and I'll do a PR.